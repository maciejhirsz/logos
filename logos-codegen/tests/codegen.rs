use std::{error::Error, io, path::PathBuf};

#[rstest::rstest]
#[case("simple")]
#[case("no_error_lut")]
pub fn test_codegen(#[case] fixture: &str) -> Result<(), Box<dyn Error>> {
    let mut fixture_dir = PathBuf::new();
    fixture_dir.push(env!("CARGO_MANIFEST_DIR"));
    fixture_dir.push("tests");
    fixture_dir.push("data");
    fixture_dir.push(fixture);

    let input = fixture_dir.join("input.rs");
    fixture_dir.push("output.rs");
    let output_file = fixture_dir;

    let input = std::fs::read_to_string(input)?;
    let output = std::fs::read_to_string(&output_file)?;

    let generated = logos_codegen::generate(input.parse()?);
    let generated = generated.to_string();

    if std::env::var("BLESS_CODEGEN").is_ok_and(|value| value == "1") {
        std::fs::write(&output_file, &generated)?;
        return Ok(());
    }

    assert_eq!(generated, output, "Codegen test failed: `{fixture}`, run tests again with env var `BLESS_CODEGEN=1` to bless these changes");

    Ok(())
}
