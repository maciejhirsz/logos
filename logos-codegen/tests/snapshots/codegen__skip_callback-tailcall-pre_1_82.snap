---
source: logos-codegen/tests/codegen.rs
expression: formatted
---
impl<'s> ::logos::Logos<'s> for Token {
    type Error = &'static str;
    type Extras = Vec<&'static str>;
    type Source = str;
    fn lex(
        lex: &mut ::logos::Lexer<'s, Self>,
    ) -> std::option::Option<std::result::Result<Self, Self::Error>> {
        use ::logos::internal::{
            LexerInternal, CallbackRetVal, CallbackResult, SkipRetVal, SkipResult,
        };
        use std::result::Result as _Result;
        use std::option::Option as _Option;
        use ::logos::Logos;
        type _Lexer<'s> = ::logos::Lexer<'s, Token>;
        macro_rules! _fast_loop {
            ($lex:ident, $test:ident, $offset:ident) => {
                'fast_loop : { while let Some(arr) = $lex .read:: < & [u8; 8usize] >
                ($offset) { if $test (arr[0usize]) { $offset += 0usize; break 'fast_loop;
                } if $test (arr[1usize]) { $offset += 1usize; break 'fast_loop; } if
                $test (arr[2usize]) { $offset += 2usize; break 'fast_loop; } if $test
                (arr[3usize]) { $offset += 3usize; break 'fast_loop; } if $test
                (arr[4usize]) { $offset += 4usize; break 'fast_loop; } if $test
                (arr[5usize]) { $offset += 5usize; break 'fast_loop; } if $test
                (arr[6usize]) { $offset += 6usize; break 'fast_loop; } if $test
                (arr[7usize]) { $offset += 7usize; break 'fast_loop; } $offset += 8usize;
                } while let Some(byte) = $lex .read:: < u8 > ($offset) { if $test (byte)
                { break 'fast_loop; } $offset += 1; } }
            };
        }
        #[inline]
        fn _make_error<'s>(lex: &mut _Lexer<'s>) -> <Token as Logos<'s>>::Error {
            <Token as Logos<'s>>::Error::default()
        }
        fn state0<'s>(
            lex: &mut _Lexer<'s>,
            mut offset: usize,
        ) -> _Option<_Result<Token, <Token as Logos<'s>>::Error>> {
            let other = lex.read::<u8>(offset);
            if let Some(byte) = other {
                enum LogosNextState {
                    state1,
                    state2,
                    state3,
                    state4,
                    state5,
                    state6,
                }
                const TABLE: [_Option<LogosNextState>; 256] = [
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    Some(LogosNextState::state1),
                    Some(LogosNextState::state2),
                    Some(LogosNextState::state3),
                    Some(LogosNextState::state4),
                    Some(LogosNextState::state5),
                    Some(LogosNextState::state5),
                    Some(LogosNextState::state6),
                    Some(LogosNextState::state6),
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                    None,
                ];
                match TABLE[byte as usize] {
                    Some(LogosNextState::state1) => {
                        offset += 1;
                        return state1(lex, offset);
                    }
                    Some(LogosNextState::state2) => {
                        offset += 1;
                        return state2(lex, offset);
                    }
                    Some(LogosNextState::state3) => {
                        offset += 1;
                        return state3(lex, offset);
                    }
                    Some(LogosNextState::state4) => {
                        offset += 1;
                        return state4(lex, offset);
                    }
                    Some(LogosNextState::state5) => {
                        offset += 1;
                        return state5(lex, offset);
                    }
                    Some(LogosNextState::state6) => {
                        offset += 1;
                        return state6(lex, offset);
                    }
                    None => {}
                }
            } else {
                if lex.offset() == offset {
                    return None;
                }
            }
            lex.end_to_boundary(offset + if other.is_some() { 1 } else { 0 });
            return Some(Err(_make_error(lex)));
        }
        fn state1<'s>(
            lex: &mut _Lexer<'s>,
            mut offset: usize,
        ) -> _Option<_Result<Token, <Token as Logos<'s>>::Error>> {
            lex.end(offset);
            let other = lex.read::<u8>(offset);
            if let Some(byte) = other {} else {}
            lex.trivia();
            offset = lex.offset();
            return state0(lex, offset);
        }
        fn state2<'s>(
            lex: &mut _Lexer<'s>,
            mut offset: usize,
        ) -> _Option<_Result<Token, <Token as Logos<'s>>::Error>> {
            lex.end(offset);
            let other = lex.read::<u8>(offset);
            if let Some(byte) = other {} else {}
            #[inline]
            fn callback<'s>(lex: &mut _Lexer<'s>) -> impl SkipRetVal<'s, Token> {
                lex.extras.push("inline_callback")
            }
            let action = SkipRetVal::<'s, Token>::construct(callback(lex));
            match action {
                SkipResult::Skip => {
                    lex.trivia();
                    offset = lex.offset();
                    return state0(lex, offset);
                }
                SkipResult::Error(err) => {
                    return Some(Err(err));
                }
            }
        }
        fn state3<'s>(
            lex: &mut _Lexer<'s>,
            mut offset: usize,
        ) -> _Option<_Result<Token, <Token as Logos<'s>>::Error>> {
            lex.end(offset);
            let other = lex.read::<u8>(offset);
            if let Some(byte) = other {} else {}
            let action = SkipRetVal::<'s, Token>::construct(labelled_callback(lex));
            match action {
                SkipResult::Skip => {
                    lex.trivia();
                    offset = lex.offset();
                    return state0(lex, offset);
                }
                SkipResult::Error(err) => {
                    return Some(Err(err));
                }
            }
        }
        fn state4<'s>(
            lex: &mut _Lexer<'s>,
            mut offset: usize,
        ) -> _Option<_Result<Token, <Token as Logos<'s>>::Error>> {
            lex.end(offset);
            let other = lex.read::<u8>(offset);
            if let Some(byte) = other {} else {}
            let action = SkipRetVal::<'s, Token>::construct(labelled_skip_callback(lex));
            match action {
                SkipResult::Skip => {
                    lex.trivia();
                    offset = lex.offset();
                    return state0(lex, offset);
                }
                SkipResult::Error(err) => {
                    return Some(Err(err));
                }
            }
        }
        fn state5<'s>(
            lex: &mut _Lexer<'s>,
            mut offset: usize,
        ) -> _Option<_Result<Token, <Token as Logos<'s>>::Error>> {
            lex.end(offset);
            let other = lex.read::<u8>(offset);
            if let Some(byte) = other {} else {}
            let action = SkipRetVal::<
                's,
                Token,
            >::construct(labelled_result_callback(lex));
            match action {
                SkipResult::Skip => {
                    lex.trivia();
                    offset = lex.offset();
                    return state0(lex, offset);
                }
                SkipResult::Error(err) => {
                    return Some(Err(err));
                }
            }
        }
        fn state6<'s>(
            lex: &mut _Lexer<'s>,
            mut offset: usize,
        ) -> _Option<_Result<Token, <Token as Logos<'s>>::Error>> {
            lex.end(offset);
            let other = lex.read::<u8>(offset);
            if let Some(byte) = other {} else {}
            let action = SkipRetVal::<
                's,
                Token,
            >::construct(labelled_skip_result_callback(lex));
            match action {
                SkipResult::Skip => {
                    lex.trivia();
                    offset = lex.offset();
                    return state0(lex, offset);
                }
                SkipResult::Error(err) => {
                    return Some(Err(err));
                }
            }
        }
        fn state7<'s>(
            lex: &mut _Lexer<'s>,
            mut offset: usize,
        ) -> _Option<_Result<Token, <Token as Logos<'s>>::Error>> {
            let other = lex.read::<u8>(offset);
            if let Some(byte) = other {} else {}
            let action = SkipRetVal::<
                's,
                Token,
            >::construct(labelled_skip_result_callback(lex));
            match action {
                SkipResult::Skip => {
                    lex.trivia();
                    offset = lex.offset();
                    return state0(lex, offset);
                }
                SkipResult::Error(err) => {
                    return Some(Err(err));
                }
            }
        }
        fn state8<'s>(
            lex: &mut _Lexer<'s>,
            mut offset: usize,
        ) -> _Option<_Result<Token, <Token as Logos<'s>>::Error>> {
            let other = lex.read::<u8>(offset);
            if let Some(byte) = other {} else {}
            let action = SkipRetVal::<
                's,
                Token,
            >::construct(labelled_result_callback(lex));
            match action {
                SkipResult::Skip => {
                    lex.trivia();
                    offset = lex.offset();
                    return state0(lex, offset);
                }
                SkipResult::Error(err) => {
                    return Some(Err(err));
                }
            }
        }
        fn state9<'s>(
            lex: &mut _Lexer<'s>,
            mut offset: usize,
        ) -> _Option<_Result<Token, <Token as Logos<'s>>::Error>> {
            let other = lex.read::<u8>(offset);
            if let Some(byte) = other {} else {}
            let action = SkipRetVal::<'s, Token>::construct(labelled_skip_callback(lex));
            match action {
                SkipResult::Skip => {
                    lex.trivia();
                    offset = lex.offset();
                    return state0(lex, offset);
                }
                SkipResult::Error(err) => {
                    return Some(Err(err));
                }
            }
        }
        fn state10<'s>(
            lex: &mut _Lexer<'s>,
            mut offset: usize,
        ) -> _Option<_Result<Token, <Token as Logos<'s>>::Error>> {
            let other = lex.read::<u8>(offset);
            if let Some(byte) = other {} else {}
            let action = SkipRetVal::<'s, Token>::construct(labelled_callback(lex));
            match action {
                SkipResult::Skip => {
                    lex.trivia();
                    offset = lex.offset();
                    return state0(lex, offset);
                }
                SkipResult::Error(err) => {
                    return Some(Err(err));
                }
            }
        }
        fn state11<'s>(
            lex: &mut _Lexer<'s>,
            mut offset: usize,
        ) -> _Option<_Result<Token, <Token as Logos<'s>>::Error>> {
            let other = lex.read::<u8>(offset);
            if let Some(byte) = other {} else {}
            #[inline]
            fn callback<'s>(lex: &mut _Lexer<'s>) -> impl SkipRetVal<'s, Token> {
                lex.extras.push("inline_callback")
            }
            let action = SkipRetVal::<'s, Token>::construct(callback(lex));
            match action {
                SkipResult::Skip => {
                    lex.trivia();
                    offset = lex.offset();
                    return state0(lex, offset);
                }
                SkipResult::Error(err) => {
                    return Some(Err(err));
                }
            }
        }
        fn state12<'s>(
            lex: &mut _Lexer<'s>,
            mut offset: usize,
        ) -> _Option<_Result<Token, <Token as Logos<'s>>::Error>> {
            let other = lex.read::<u8>(offset);
            if let Some(byte) = other {} else {}
            lex.trivia();
            offset = lex.offset();
            return state0(lex, offset);
        }
        state0(lex, lex.offset())
    }
}
